### PR TITLE
Auro-Modal Initial Implementation

### PR DESCRIPTION
**Pull request by gusnaughton**
# Alaska Airlines Pull Request

As a user, I want to see a modal in both blocking and non-blocking instances based on context. 

Blocking modals current rely on the implementing team to create a call to action that removes the `open` attribute from the modal element to trigger the close.

## Type of change:

Please delete options that are not relevant.

- [x] New capability

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

**Comment by gusnaughton** [comment]
CSSLint fails due to the transition being a duplicate element.

File: src/auro-modal.js [hunk 1]
```diff
@@ -8,27 +8,39 @@ import { LitElement, html, css } from "lit-element";
...
   static get properties() {
     return {
-      cssClass:   { type: String }
+      header: { type: String },
+      blocking: { type: Boolean },
+      open:   { type: Boolean }
     };
   }
 
+  toggleViewable(evt) {
```
Comment (ecommerceUXdesign) [inline src/auro-modal.js:39]: Please add jSDoc comments for this function

**Comment by ecommerceUXdesign** [comment]
What is the trigger for closing a 'blocking' modal? What event will a user use for this?

**Comment by gusnaughton** [comment]
> What is the trigger for closing a 'blocking' modal? What event will a user use for this?

The implementation would need to provide a function to remove the `open` attribute from the element, similar to our [saver-to-main-upsell-bottom](https://components.alaskaair.com/demo/saver-to-main-upsell-bottom). We don't have a guarantee that the call to action would be a button, it could be an image.

**Comment by blackfalcon** [comment]
> > What is the trigger for closing a 'blocking' modal? What event will a user use for this?
> 
> The implementation would need to provide a function to remove the `open` attribute from the element, similar to our [saver-to-main-upsell-bottom](https://components.alaskaair.com/demo/saver-to-main-upsell-bottom). We don't have a guarantee that the call to action would be a button, it could be an image.

I would suggest working this action into the demo and we need this in the documentation. We can't assume that devs will know right away how this works leading to confusion and unnecessary wrappers.

**Review by ecommerceUXdesign — CHANGES_REQUESTED** [review]

**Comment by blackfalcon** [comment]
What's the status of this PR @gusnaughton ?

**Comment by gusnaughton** [comment]
Based on this feedback, this branch will need a total refactor. I am working on this as my top priority.

**Comment by gusnaughton** [comment]
Implemented feedback. Looking for suggestions on how to handle intercepting the click on the modal's pane when it appears. Thanks for the insightful feedback, this scss structure looks a lot better!

**Comment by gusnaughton** [comment]
Bug: Positioning of the top placement of the modal in desktop view depends on content size.
Bug: Clicking on the non-blocking modal closes it, instead of the close button or background.

Would appreciate some advice on that.

**Comment by gusnaughton** [comment]
Bumping for review

**Comment by blackfalcon** [comment]
> Bug: Positioning of the top placement of the modal in desktop view depends on content size.
> Bug: Clicking on the non-blocking modal closes it, instead of the close button or background.
> 
> Would appreciate some advice on that.

I made a commit that suggests some updates. I can't find the comp, it's not linked to any of the issues, so I just roughed it. So, top placement, I took a guess. As for the clicking of the non-blocking, you have the event on the outermost div, so that whole area is clickable? 

I would suggest that the overlay and the modal are two different things. The modal will be absolutely positioned, so just put it on top and let the overlay be in the background. Then you can click on the overlay and the modal itself will not be clickable.

File: docs/api.md [hunk 2]
```diff
@@ -4,6 +4,22 @@ auro-modal provides users a way to ...
...
-| Property   | Attribute  | Type     | Description                                  |
-|------------|------------|----------|----------------------------------------------|
-| `cssClass` | `cssClass` | `String` | Applies designated CSS class to DOM element. |
+| Property   | Attribute  | Type         | Default      | Description                                      |
+|------------|------------|--------------|--------------|--------------------------------------------------|
+| `blocking` | `blocking` | `Boolean`    |              | Blocking modals force the user to take an action (no close button) |
+| `dom`      |            |              |              |                                                  |
+| `header`   | `header`   | `String`     |              | Text to display as the header of the modal.      |
+| `open`     | `open`     | `open`       |              | this attr forces the modal to open               |
+| `svg`      |            | `any \| any` | "firstChild" |                                                  |
```
Comment (blackfalcon) [inline docs/api.md:13]: needs description

File: test/auro-modal.test.js [hunk 3]
```diff
@@ -25,4 +15,46 @@ describe('auro-modal', () => {
...
+  it('auro-modal non-blocking renders a close icon', async () => {
+    const el = await fixture(html`
+      <auro-modal></auro-modal>
+    `);
+
+    const root = el.shadowRoot;
+    const title = root.querySelector('#modal-close');
+    await expect(title).to.not.equal(null);
+  });
+  // it('auro-modal renders a header', async () => {
```
Comment (blackfalcon) [inline test/auro-modal.test.js:39]: I broke this test ... sorry, not sure what I did O_O

File: src/auro-modal.js [hunk 4]
```diff
@@ -37,9 +52,38 @@ class AuroModal extends LitElement {
...
+      'util_insetXxxl--squish': true,
+      'modal': true,
+      'modal--open': this.open,
+    }
+
+
     return html`
-      <div class=${this.cssClass}>
-        <slot></slot>
+      <div class="${classMap(classes)}" id="modal-overlay" @click=${this.blocking ? null : this.toggleViewable}>
```
Comment (blackfalcon) [inline src/auro-modal.js:70]: Click event attached to parent div. Break into to individual things so that you don't have this outer relationship.

File: src/auro-modal.js [hunk 5]
```diff
@@ -37,9 +52,38 @@ class AuroModal extends LitElement {
...
+        <div class="${classMap(contentClasses)}">
+          <div class="modalHeader">
+            <div class="modalHeader--heading" id="modal-header">
```
Comment (blackfalcon) [inline src/auro-modal.js:73]: I'd prefer that this was the header component or a [pre-defined header CSS selector from WCSS](https://auro.alaskaair.com/typography/usage). By doing this it is a unique flower that needs to be updated if there are any changes.

File: src/auro-modal.js [hunk 6]
```diff
@@ -4,31 +4,46 @@
...
+    this.dom = new DOMParser().parseFromString(closeIcon.svg, 'text/html');
+    this.svg = this.dom.body.firstChild;
+  }
 
-  // function to define props used within the scope of thie component
+  // function to define props used within the scope of this component
   static get properties() {
     return {
-      cssClass:   { type: String }
+      header: { type: String },
```
Comment (blackfalcon) [inline src/auro-modal.js:36]: I'd prefer that this was using a slot versus a prop. Mor reflective of HTML use versus making complex APIs.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

**Comment by gusnaughton** [comment]
Your changes look great, however I'm not sure if we can use `top: 0` as the final solution. [The mockups require the modal to fit to size](https://share.goabstract.com/af683106-5138-4d02-8982-006211aeb920?collectionLayerId=28e70bcf-37e2-481e-9bd2-ecafcd101c3e&mode=design). Is there another way we can approach this?

File: docs/api.md (hunk 2 above)
Comment (gusnaughton) [reply docs/api.md:13]: svg is an internal variable. Can we omit it?

File: src/auro-modal.js (hunk 6 above)
Comment (gusnaughton) [reply src/auro-modal.js:36]: While I generally agree with you on this, I'm wary of making this a slot because it will lead to inconsistent header implementations in modals down the line. @Crea-Ive @vidalmenAS - do we want flexibility in how a modal's header may be displayed? Or should we build this to ensure that the header will ONLY ever look like this text style:
image

File: src/auro-modal.js (hunk 5 above)
Comment (gusnaughton) [reply src/auro-modal.js]: Good point. Pending below discussion regarding slots/props.

File: test/auro-modal.test.js (hunk 3 above)
Comment (gusnaughton) [reply test/auro-modal.test.js:39]: pending below discussion about slots/props

File: docs/api.md (hunk 2 above)
Comment (blackfalcon) [reply docs/api.md:13]: No, so say `internal ...` in the description.

File: src/auro-modal.js (hunk 6 above)
Comment (blackfalcon) [reply src/auro-modal.js:37]: I was not suggesting flexibility. You can use the header WC with a fixed setting and just update the string. Or, you can use the predefined CSS on the element you already created. 

My goal is that you are not re-doing the CSS values on the selector.

**Comment by blackfalcon** [comment]
> Your changes look great, however I'm not sure if we can use `top: 0` as the final solution. [The mockups require the modal to fit to size](https://share.goabstract.com/af683106-5138-4d02-8982-006211aeb920?collectionLayerId=28e70bcf-37e2-481e-9bd2-ecafcd101c3e&mode=design). Is there another way we can approach this?

My suggestion was not to use `top: 0` as the end result for the UI. The question was to get it to go to the top, so I did that. As stated, I could not find the comp, so I didn't know what the end design was supposed to be. 

I think from the example code update I made, you can make it work as the key was moving around some CSS properties on how the selectors were being used. If you are still stuck, let me know and I will take another crack at it.

**Comment by gusnaughton** [comment]
> > Your changes look great, however I'm not sure if we can use `top: 0` as the end result. [The mockups require the modal to fit to size](https://share.goabstract.com/af683106-5138-4d02-8982-006211aeb920?collectionLayerId=28e70bcf-37e2-481e-9bd2-ecafcd101c3e&mode=design). Is there another way we can approach this?
> 
> My suggestion was not to use `top: 0` as the end result for the UI. The question was to get it to go to the top, so I did that. As stated, I could not find the comp, so I didn't know what the end design was supposed to be.
> 
> I think from the example code update I made, you can make it work as the key was moving around some CSS properties on how the selectors were being used. If you are still stuck, let me know and I will take another crack at it.

This makes sense. 

Based on this mockup, we need to figure out a way to make the modal appear in the center of the screen, fit to content, regardless of where the scrollbar is on the page (the modal should always be right in the center):

image

This is why I was toying with the transform: translateY() property, but obviously that didn't work well either.

**Comment by blackfalcon** [comment]
@gusnaughton If you are asking for my assistance to complete this, let me know. I am not sure what the next step is?

File: src/auro-modal.js (hunk 6 above)
Comment (gusnaughton) [reply src/auro-modal.js:37]: Fixed!

File: src/auro-modal.js (hunk 5 above)
Comment (gusnaughton) [reply src/auro-modal.js]: Replaced with the auro docs, but I had to unset `margin-block-end` and `margin-block-start`.

**Comment by gusnaughton** [comment]
> @gusnaughton If you are asking for my assistance to complete this, let me know. I am not sure what the next step is?

Apologies, yesterday I had 8.5 hours of back-to-back meetings, so I could not work on this.

I have updated the header mechanism, tests, and styles to remove the margin from the header. The next step is to fix the sizing of the modal to be like the screenshot I pasted above, and that's where I'm struggling. Help with this would be greatly appreciated. Apart from that, I believe we're in good shape.

File: src/auro-modal.js [hunk 7]
```diff
@@ -37,9 +52,38 @@ class AuroModal extends LitElement {
...
+      </div>
+      <div class="${classMap(contentClasses)}">
+        <div class="modalHeader">
+          <h1 class="auro_heading auro_heading--700 modalHeader--noMargin">
+            <slot name="modal-header"></slot>
+          </h1>
+          ${this.blocking
+            ? html``
+            : html`
+              <div class="modalHeader--action" id="modal-close" @click="${this.toggleViewable}">
```
Comment (geoffrich) [inline src/auro-modal.js:80]: This needs to be changed to a `button` for accessibility. [Divs with an onclick are not accessible](https://adrianroselli.com/2016/01/links-buttons-submits-and-divs-oh-hell.html#WorstPractice) -- they are not recognized as buttons by screen readers and cannot be navigated to by keyboard users. This means a keyboard user would be unable to close this modal. 

The button also needs [an accessible name](https://www.sarasoueidan.com/blog/accessible-icon-buttons/#technique-1-accessible-visually-hidden-text). Otherwise, screen reader users won't know the purpose of the button.

File: src/auro-modal.js [hunk 8]
```diff
@@ -4,31 +4,46 @@
...
+  // function to define props used within the scope of this component
   static get properties() {
     return {
-      cssClass:   { type: String }
+      blocking: { type: Boolean },
+      open:   { type: Boolean }
     };
   }
 
+  toggleViewable(evt) {
```
Comment (geoffrich) [inline src/auro-modal.js:42]: We should be implementing the [ARIA authoring practices for modal](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). In summary:

1. When opening the modal, focus should be moved into an element within the modal
2. When closing the modal, focus should return to the element that opened the modal or another logical element.
3. The escape key closes the modal.
4. Focus is trapped -- keyboard users cannot tab to elements outside of the modal.
3. Modal container has `role="dialog"` 
4. The modal container has a label via `aria-labelledby` or `aria-label`.

The spec also suggests using `aria-modal="true"` on the container to prevent screen readers from discovering content outside the modal, but [that seems to have issues in VoiceOver/Safari](https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/). We might need to apply "aria-hidden" and inert instead. See the heading "Updates to ARIA and using inert" in the previous link.

These suggestions are all important to ensure that keyboard and screen reader users have the same experience as everyone else. Without them, it is not clear that a modal opened or how to interact with it.

**Review by geoffrich — CHANGES_REQUESTED** [review]
I added some comments around accessibility. Currently the modal is not accessible to keyboard or screen reader users. Let me know if you have any questions about implementing these.

**Comment by gusnaughton** [comment]
> I added some comments around accessibility. Currently the modal is not accessible to keyboard or screen reader users. Let me know if you have any questions about implementing these.

Thanks for this analysis, this helps me a lot! I'll get to work implementing this.

File: src/auro-modal.js (hunk 7 above)
Comment (blackfalcon) [reply src/auro-modal.js:80]: The article you linked, while correct, I feel that it is a hostile opinion. The various tools are made available to address situations when needed. I am not going to 'slap' someone if they don't use a button. 

I am of the opinion, and the spec is, the user is made aware of the action and ability to interact. 

In regards to the hidden text, that is correct. There are a couple of things here. One, you can put text inside the button and use this selector https://alaskaairlines.github.io/WebCoreStyleSheets/#utility-display-css-#{$scope}.#{$prefix}util_displayHiddenVisually

The X SVG itself has a title, but it's currently only descriptive of the image, not the action.

Comment (gusnaughton) [reply src/auro-modal.js:80]: In terms of button, I'm curious about what your recommended path forward is here @blackfalcon. My instinct was to use the `<auro-button-ld tertiary />` component, but there's too much left/right padding - if we move forward with this, it sounds like we need to create IconButton.

Comment (geoffrich) [reply src/auro-modal.js:80]: Apologies for coming off as hostile! In hindsight, I should have linked something that wasn't so aggressive. 

IconButton could be a good candidate for a standard/design component in the future (including button styles reset, focus styles, accessible name). I think we could also get away with keeping everything local to this component in the short term and break it out later, especially if we don't have a standard design for it yet.

Comment (gusnaughton) [reply src/auro-modal.js:80]: No hostility detected. @blackfalcon, how would you prefer we move forward?

File: src/auro-modal.js (hunk 8 above)
Comment (gusnaughton) [reply src/auro-modal.js:42]: For my own notes:

- [ ] When opening the modal, focus should be moved into an element within the modal

- [ ] When closing the modal, focus should return to the element that opened the modal or another logical element.

- [ ] The escape key closes the modal.

- [ ] Focus is trapped -- keyboard users cannot tab to elements outside of the modal (this is intended for blocking modals).

- [ ] Modal container has role="dialog"

- [ ] The modal container has a label via aria-labelledby or aria-label

**Comment by gusnaughton** [comment]
We had some internal discussions regarding accessibility.

Are we okay with shipping v1 with the content escaping accessibility issue? If so, the only item pending on this list is the content-box sizing as per the mockups, which I need help with implementing.

**Comment by geoffrich** [comment]
> Are we okay with shipping v1 with the content escaping accessibility issue? If so, the only item pending on this list is the content-box sizing as per the mockups, which I need help with implementing.

Not sure what you are referring to as the "content escaping accessibility issue." Is that hiding non-modal content from screen readers? If so, we can probably wait to fix that issue until later. Let's make sure an issue is created to track it.

IMO, I don't think we can ship a v1 without end-to-end keyboard accessibility. In essence, this includes:
1. _When opening the modal with the keyboard, my focus is moved inside the modal._ This can be accomplished by calling `.focus()` on an interactive element inside the modal. Without this, it is hard for me to interact with the modal because I might have to tab through many other elements to get there.
2. _I can close the modal with the keyboard. When I do so, my focus is returned to where I was when I opened the modal._ This can be accomplished by storing document.activeElement when the modal is opened and calling `.focus()` on it when the modal is closed. Without this, my focus position is lost and I have to navigate back from the start.

Without those two items, it will be very difficult for keyboard users to find and interact with the modal. Focus trapping is also important, but we could create an issue for it if you don't have time to implement it. Without focus trapping, keyboard users could accidentally escape the modal, but at least they would be placed in the modal to begin with.

In summary, at minimum we need to manage focus properly when the modal is opened and closed (items 1 and 2). Everything else I listed in my previous comment is still important and should be created as high-priority issues if we don't have time to implement for initial release. However, in the end it's @blackfalcon's call as to what he's comfortable shipping as v1.

**Review by ecommerceUXdesign — APPROVED** [review]

**Review by ecommerceUXdesign — APPROVED** [review]
